src/Microsoft.NET.Sdk.Functions.MSBuild/Tasks/ZipDeploymentStatus.cs

```diff
@@ -0,0 +1,127 @@
+{
+    internal class ZipDeploymentStatus
+    {
+        private const int MaxMinutesToWait = 3;
+        private const int StatusRefreshDelaySeconds = 3;
+        private const int RetryCount = 3;
+        private const int RetryDelaySeconds = 1;
+
+        private readonly IHttpClient _client;
+        private readonly string _userAgent;
+        private readonly TaskLoggingHelper _log;
+        private readonly bool _logMessages;
+
+        public ZipDeploymentStatus(IHttpClient client, string userAgent, TaskLoggingHelper log, bool logMessages)
+        {
+            _client = client;
+            _userAgent = userAgent;
+            _log = log;
+            _logMessages = logMessages;
+        }
+
+        public async Task<DeployStatus> PollDeploymentStatus(string deploymentUrl, string userName, string password)
+        {
+            DeployStatus deployStatus = DeployStatus.Pending;
+            var tokenSource = new CancellationTokenSource(TimeSpan.FromMinutes(MaxMinutesToWait));
+            if (_logMessages)
+            {
+                _log.LogMessage(Resources.DeploymentStatusPolling);
+            }
+            while (!tokenSource.IsCancellationRequested && deployStatus != DeployStatus.Success && deployStatus != DeployStatus.Failed && deployStatus != DeployStatus.Unknown)
+            {
+                try
+                {
+                    deployStatus = await GetDeploymentStatus(deploymentUrl, userName, password, RetryCount, TimeSpan.FromSeconds(RetryDelaySeconds));
+                    if (_logMessages)
+                    {
+                        _log.LogMessage(String.Format(Resources.DeploymentStatus, Enum.GetName(typeof(DeployStatus), deployStatus)));
+                    }
+                }
+                catch (HttpRequestException)
+                {
+                    return DeployStatus.Unknown;
+                }
+
+                await Task.Delay(TimeSpan.FromSeconds(StatusRefreshDelaySeconds));
+            }
+
+            return deployStatus;
+        }
+
+        private async Task<DeployStatus> GetDeploymentStatus(string deploymentUrl, string userName, string password, int retryCount, TimeSpan retryDelay)
+        {
+            Dictionary<string, string> json = await InvokeGetRequestWithRetry<Dictionary<string, string>>(deploymentUrl, userName, password, retryCount, retryDelay);
+            string statusString = null;
+            if (json!= null && !json.TryGetValue("status", out statusString))
+            {
+                return DeployStatus.Unknown;
+            }
+
+            if (Enum.TryParse(statusString, out DeployStatus result))
+            {
+                return result;
+            }
+
+            return DeployStatus.Unknown;
+        }
+
+        private async Task<T> InvokeGetRequestWithRetry<T>(string url, string userName, string password, int retryCount, TimeSpan retryDelay)
+        {
+            IHttpResponse response = null;
+            await Retry(async () =>
```

Comment:
RetryAsync?  Should this method and RetryAsync take the cancellationToken for the caller (which intends to cancel after a certain duration)?

Reply:
Changed to RetryAsync. I don't think we need to flow the token to retry but will flow the token to getAsync call itself.